tests/unit/core/test_nemo_callbacks.py

```diff
@@ -0,0 +1,212 @@
+                optimizer="sgd",
+            )
+
+        output_lines = std_out.getvalue().splitlines()
+        assert len(output_lines) == 4
+        for line in output_lines:
+            assert "y_pred" in line
+
+    @pytest.mark.unit
+    def test_TensorboardLogger(self, clean_up, create_tensorboard_file):
+        data_source = RealFunctionDataLayer(n=100, batch_size=1)
+        trainable_module = TaylorNet(dim=4)
+        loss = MSELoss()
+
+        # Create the graph by connnecting the modules.
+        x, y = data_source()
+        y_pred = trainable_module(x=x)
+        loss_tensor = loss(predictions=y_pred, target=y)
+
+        tb_logger = TensorboardLogger(create_tensorboard_file, step_freq=1)
+        callbacks = [tb_logger]
+
+        self.nf.train(
+            tensors_to_optimize=[loss_tensor],
+            callbacks=callbacks,
+            optimization_params={"max_steps": 4, "lr": 0.01},
+            optimizer="sgd",
+        )
+
+        # efi.inspect("temp", tag="loss")
+        inspection_units = efi.get_inspection_units("temp", "", "loss")
+
+        # Make sure there is only 1 tensorboard file
+        assert len(inspection_units) == 1
+
+        # Assert that there the loss scalars has been logged 4 times
+        assert len(inspection_units[0].field_to_obs['scalars']) == 4
+
+    @pytest.mark.unit
+    def test_epoch_decorators(self, clean_up):
+        data_source = RealFunctionDataLayer(n=24, batch_size=12)
+        trainable_module = TaylorNet(dim=4)
+        loss = MSELoss()
+
+        # Create the graph by connnecting the modules.
+        x, y = data_source()
+        y_pred = trainable_module(x=x)
+        loss_tensor = loss(predictions=y_pred, target=y)
+
+        epoch_start_counter = [0]
+        epoch_end_counter = [0]
+
+        @on_epoch_start
+        def count_epoch_starts(state, counter=epoch_start_counter):
+            counter[0] += 1
+
+        @on_epoch_end
+        def count_epoch_ends(state, counter=epoch_end_counter):
+            counter[0] -= 1
+
+        callbacks = [count_epoch_starts, count_epoch_ends]
+
+        self.nf.train(
+            tensors_to_optimize=[loss_tensor],
+            callbacks=callbacks,
+            optimization_params={"max_steps": 4, "lr": 0.01},
+            optimizer="sgd",
+        )
+
+        assert epoch_start_counter[0] == 2
+        assert epoch_end_counter[0] == -2
+
+    @pytest.mark.unit
+    def test_step_batch_decorators(self, clean_up):
+        """Showcase the difference between step and batch"""
+        data_source = RealFunctionDataLayer(n=24, batch_size=12)
+        trainable_module = TaylorNet(dim=4)
+        loss = MSELoss()
+
+        # Create the graph by connnecting the modules.
+        x, y = data_source()
+        y_pred = trainable_module(x=x)
+        loss_tensor = loss(predictions=y_pred, target=y)
+
+        epoch_step_counter = [0]
+        epoch_batch_counter = [0]
+
+        @on_step_end
```

Comment:
wow! Ok, those decorators are nice!

I haven't got that from the NeMoCallbacks API... what those are good for?